### PR TITLE
Added more tracing to driver

### PR DESCRIPTION
**Pull request by mitchell-elholm**
### Summary

AD-249 Added tracing to driver

### Description

WIP

Added tracing for:

- Translated MQL
- Time taken to translate MQL
- Time to execute query
- Metadata retrieval time
- Metadata generation time
- Metadata table addition
- Metadata conflict resolution
- Connection properties
- Logs each stage of aggregation pipeline

### Additional Reviewers
@birschick-bq
@mitchell-elholm
@alexey-temnikov 
@andiem-bq

### CODE REVIEW

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java [hunk 1]
```diff
@@ -103,17 +104,31 @@ public static DocumentDbSchema get(
             case VERSION_LATEST_OR_NEW:
                 // If latest exist, return it.
                 if (schema != null) {
+                    LOGGER.info("Attempting to retrieve latest latest metadata.");
```
Comment (andiemontoyeah) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java]: latest latest -> latest ?

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbTableSchemaGenerator.java [hunk 2]
```diff
@@ -216,6 +226,9 @@ private static void processDocument(
                 .collectionName(collectionName)
                 .columns(columnMap)
                 .build();
+        if (LOGGER.isDebugEnabled() && !tableMap.containsKey(metadataTable.getSqlName())) {
+            LOGGER.debug(String.format("2Added schema for table %s.", metadataTable.getSqlName()));
```
Comment (andiemontoyeah) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbTableSchemaGenerator.java]: 2Added -> Added?

File: .github/badges/jacoco.svg [hunk 3]
```diff
@@ -1 +1 @@
...
\ No newline at end of file
+<svg xmlns="http://www.w3.org/2000/svg" width="104" height="20" role="img" aria-label="coverage: 81.8%"><linearGradient id="s" x2="0" y2="100%"><stop offset="0" stop-color="#bbb" stop-opacity=".1"/><stop offset="1" stop-opacity=".1"/></linearGradient><clipPath id="r"><rect width="104" height="20" rx="3" fill="#fff"/></clipPath><g clip-path="url(#r)"><rect width="61" height="20" fill="#555"/><rect x="61" width="43" height="20" fill="#a4a61d"/><rect width="104" height="20" fill="url(#s)"/></g><g fill="#fff" text-anchor="middle" font-family="Verdana,Geneva,DejaVu Sans,sans-serif" text-rendering="geometricPrecision" font-size="110"><text aria-hidden="true" x="315" y="150" fill="#010101" fill-opacity=".3" transform="scale(.1)" textLength="510">coverage</text><text x="315" y="140" transform="scale(.1)" fill="#fff" textLength="510">coverage</text><text aria-hidden="true" x="815" y="150" fill="#010101" fill-opacity=".3" transform="scale(.1)" textLength="330">81.8%</text><text x="815" y="140" transform="scale(.1)" fill="#fff" textLength="330">81.8%</text></g></svg>
```
Comment (alexey-temnikov) [inline .github/badges/jacoco.svg]: Why code coverage is reduced?

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java [hunk 4]
```diff
@@ -176,6 +181,7 @@ public DocumentDbAggregate(final RelOptCluster cluster, final RelTraitSet traitS
         implementor.setMetadataTable(metadata);
         implementor.setDocumentDbTable(
                 new DocumentDbTable(implementor.getDocumentDbTable().getCollectionName(), metadata));
+        LOGGER.info("Created aggregation stages of pipeline.");
```
Comment (alexey-temnikov) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java:184]: Does this log useful without any debug information about the query / pipeline / UUID of the query? Same applicable for log statements below.

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbSort.java [hunk 5]
```diff
@@ -98,6 +104,7 @@ public DocumentDbSort(final RelOptCluster cluster, final RelTraitSet traitSet,
             implementor.add(null,
                     "{$limit: " + ((RexLiteral) fetch).getValue() + "}");
         }
+        LOGGER.info("Created sort and row limit stages of pipeline.");
```
Comment (alexey-temnikov) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbSort.java:107]: Same as above, shall we print the row limit?

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java [hunk 6]
```diff
@@ -91,33 +91,54 @@ public static DocumentDbSchema get(
...
+        final int lookupVersion = Math.max(schemaVersion, VERSION_LATEST_OR_NEW);
         // Get the latest or specific version, might not exist
-        DocumentDbSchema schema = schemaReader.read(schemaName, lookupVersion);
+        final DocumentDbSchema schema = schemaReader.read(schemaName, lookupVersion);
 
         switch (schemaVersion) {
             case VERSION_LATEST_OR_NEW:
                 // If latest exist, return it.
                 if (schema != null) {
+                    LOGGER.info("Attempting to retrieve latest metadata.");
```
Comment (alexey-temnikov) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java]: Seems like this comment is not useful, as it is immediately followed by other one which contains technical information.

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java [hunk 7]
```diff
@@ -91,33 +91,54 @@ public static DocumentDbSchema get(
...
+                    LOGGER.info(String.format("Successfully retrieved metadata in %d ms.",
+                            Instant.now().toEpochMilli() - beginRetrieval.toEpochMilli()));
                     return schema;
                 }
                 // Otherwise, fall through to create new.
+                LOGGER.info("Existing metadata not found, will generate new metadata instead.");
```
Comment (alexey-temnikov) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java]: Do we need to log database name and parameters (e.g. what metadata was specified in connection strings?)

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java [hunk 8]
```diff
@@ -91,33 +91,54 @@ public static DocumentDbSchema get(
...
                 return schema;
         }
     }
 
+    private static DocumentDbSchema getNewSchema(final DocumentDbConnectionProperties properties,
+                                                 final String schemaName,
+                                                 final Map<String, DocumentDbSchemaTable> tableMap,
+                                                 final int versionNumber) throws SQLException {
+        // Get a new version.
+        LOGGER.info("Beginning generation of new metadata.");
```
Comment (alexey-temnikov) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java]: Doesn't seem like useful information in INFO log level.

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbTableSchemaGenerator.java [hunk 9]
```diff
@@ -173,6 +176,13 @@ private static void processDocument(
             // in the root document.
             final JdbcType prevSqlType = getPrevSqlTypeOrDefault(prevMetadataColumn);
             final JdbcType nextSqlType = getSqlTypeIfIsPrimaryKey(bsonType, prevSqlType, isPrimaryKey);
+            if (LOGGER.isDebugEnabled()) {
+                final JdbcType currentDocType = getSqlTypeIfIsPrimaryKey(bsonType, JdbcType.NULL, isPrimaryKey);
+                if (!prevSqlType.equals(currentDocType) && prevMetadataColumn != null) {
+                    LOGGER.debug(String.format("Type conflict in table %s, types %s and %s mapped to %s.",
```
Comment (alexey-temnikov) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbTableSchemaGenerator.java:182]: Should conflicts be logged with Info level?

File: src/main/java/software/amazon/documentdb/jdbc/DocumentDbConnection.java [hunk 10]
```diff
@@ -61,6 +61,17 @@
             throws SQLException {
         super(connectionProperties);
         this.connectionProperties = connectionProperties;
+        if (LOGGER.isDebugEnabled()) {
+            final StringBuilder sb = new StringBuilder();
+            sb.append("Creating connection with following properties:");
+            for (String propertyName : connectionProperties.stringPropertyNames()) {
+                if (!propertyName.equals(DocumentDbConnectionProperty.PASSWORD.getName())) {
```
Comment (alexey-temnikov) [inline src/main/java/software/amazon/documentdb/jdbc/DocumentDbConnection.java]: Use constant.equals(variable) instead of variable.equals(constant)

File: src/main/java/software/amazon/documentdb/jdbc/DocumentDbQueryExecutor.java [hunk 11]
```diff
@@ -182,6 +187,10 @@ protected void cancelQuery() throws SQLException {
 
         final ImmutableList<JdbcColumnMetaData> columnMetaData = ImmutableList
                 .copyOf(queryContext.getColumnMetaData());
+        LOGGER.info(String.format("Took %d ms to execute query.",
```
Comment (alexey-temnikov) [inline src/main/java/software/amazon/documentdb/jdbc/DocumentDbQueryExecutor.java]: Time without the actual query doesn't represent useful information.

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java (hunk 4 above)
Comment (mitchell-elholm) [reply calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java:184]: My reasoning for why these are useful is so that we can see more easily where things happened if something goes wrong. It allows us to see both the timing of when each phase started (since all logs include the timestamp) as well as during which stage of translation an error occurred. The MongoBI logs did something quite similar. If this doesn't make sense they can be removed, but I think they could be useful depending on what we are trying to debug.

Edit: It might be useful to add more information, including details you mentioned.  I'll look into that.

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbTableSchemaGenerator.java (hunk 9 above)
Comment (mitchell-elholm) [reply calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbTableSchemaGenerator.java:182]: This can produce a lot of log messages which are only helpful if debugging metadata, and could hide more important log messages.

File: src/main/java/software/amazon/documentdb/jdbc/DocumentDbQueryExecutor.java (hunk 11 above)
Comment (mitchell-elholm) [reply src/main/java/software/amazon/documentdb/jdbc/DocumentDbQueryExecutor.java]: I'm not sure I agree, while we would ideally be able to log the query itself it still seems useful to log the query time. The user (including us) may know what query was being executed, and if the query time is particularly large that would indicate we should investigate further. If this is only useful with the query attached, I could change it to debug level instead.

File: .github/badges/jacoco.svg (hunk 3 above)
Comment (mitchell-elholm) [reply .github/badges/jacoco.svg]: Code coverage is decreased because none of our logging is tested, and this adds more.

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbSort.java (hunk 5 above)
Comment (mitchell-elholm) [reply calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbSort.java:107]: Yes, that's a good idea.

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java (hunk 4 above)
Comment (alexey-temnikov) [reply calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java:184]: You are right, let's see what other information can be safely added.

File: src/main/java/software/amazon/documentdb/jdbc/DocumentDbQueryExecutor.java [hunk 12]
```diff
@@ -157,12 +158,16 @@ protected void cancelQuery() throws SQLException {
      */
     @VisibleForTesting
     protected java.sql.ResultSet runQuery(final String sql) throws SQLException {
+        final Instant beginTranslation = Instant.now();
+        LOGGER.debug(String.format("SQL query: %s", sql));
+        LOGGER.info("Begin translating query.");
```
Comment (alexey-temnikov) [inline src/main/java/software/amazon/documentdb/jdbc/DocumentDbQueryExecutor.java]: Should we log query UUID across all logs, so it is easier to match the logs (especially if multiple queries are executed in parallel)?

File: src/main/java/software/amazon/documentdb/jdbc/DocumentDbQueryExecutor.java (hunk 11 above)
Comment (alexey-temnikov) [reply src/main/java/software/amazon/documentdb/jdbc/DocumentDbQueryExecutor.java]: Suggesting to add query UUID, so we can correlate all logs, including this one.

File: .github/badges/jacoco.svg (hunk 3 above)
Comment (alexey-temnikov) [reply .github/badges/jacoco.svg]: While we do not have assertions to test the logs, during the integration tests we should be creating the logs, and thus it should be covered. Perhaps we should add few test runs with debug log level, since there is minimum business logic added in the if(debug) blocks, not only log statements.

**Review by andiemontoyeah — APPROVED** [review]
lgtm

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java [hunk 13]
```diff
@@ -91,33 +91,54 @@ public static DocumentDbSchema get(
...
         switch (schemaVersion) {
             case VERSION_LATEST_OR_NEW:
                 // If latest exist, return it.
                 if (schema != null) {
+                    LOGGER.info(String.format("Successfully retrieved metadata schema %s in %d ms.",
+                            schemaName, Instant.now().toEpochMilli() - beginRetrieval.toEpochMilli()));
                     return schema;
                 }
                 // Otherwise, fall through to create new.
+                LOGGER.info(String.format("Existing metadata not found for schema %s, will generate new metadata instead for database %s.",
```
Comment (birschick-bq) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java:111]: Remove comment "... fall through to ..."

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java [hunk 14]
```diff
@@ -91,33 +91,54 @@ public static DocumentDbSchema get(
...
-                return schema;
+                final int newVersionNumber = schema != null ? schema.getSchemaVersion() + 1 : 1;
+                return getNewSchema(properties, schemaName, tableMap, newVersionNumber);
             case VERSION_LATEST_OR_NONE:
             default:
                 // Return specific version or null.
+                if (schema != null) {
+                    LOGGER.info(String.format("Retrieved schema %s version %d in %d ms.",
+                            schemaName, schemaVersion, Instant.now().toEpochMilli() - beginRetrieval.toEpochMilli()));
+                }
```
Comment (birschick-bq) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java:125]: Maybe add `else` for schema not found?

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java [hunk 15]
```diff
@@ -91,33 +91,54 @@ public static DocumentDbSchema get(
...
                     return schema;
                 }
                 // Otherwise, fall through to create new.
+                LOGGER.info(String.format("Existing metadata not found for schema %s, will generate new metadata instead for database %s.",
+                        schemaName, properties.getDatabase()));
+                return getNewSchema(properties, schemaName, tableMap, 1);
             case VERSION_NEW:
-                // Get a new version.
-                lookupVersion = schema != null ? schema.getSchemaVersion() + 1 : 1;
-                schema = getNewDatabaseMetadata(properties, schemaName, lookupVersion, tableMap);
```
Comment (birschick-bq) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java:112]: Is `getNewDatabaseMetadata` used anymore?

**Review by birschick-bq — COMMENTED** [review]
Small changes...

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java (hunk 15 above)
Comment (mitchell-elholm) [reply calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/metadata/DocumentDbMetadataService.java:112]: Not really, refactored to directly call getNewSchema, and then added the logging to that method, which I think is more clear and makes more sense.

**Review by birschick-bq — APPROVED** [review]
Great work!
